twoliter/src/lock.rs

```diff
@@ -139,15 +249,181 @@ impl Lock {
+                    digest_uri, oci_archive_str
+                )
+            );
+        }
+        let mut oci_file = File::open(&oci_archive_path).context(format!(
+            "failed to open oci archive at {}",
+            oci_archive_path.display()
+        ))?;
+        let mut oci_archive = Archive::new(&mut oci_file);
+        oci_archive
+            .unpack(path)
+            .context("failed to unpack oci image")?;
+        let index_bytes = read(path.join("index.json")).await?;
+        let index = serde_json::from_slice(index_bytes.as_slice())
+            .context("failed to deserialize oci image index")?;
+        write(&digest_file, digest).await.context(format!(
+            "failed to record digest to {}",
+            digest_file.display()
+        ))?;
+        Ok(index)
+    }
+
+    async fn extract_kit<P>(
+        &self,
+        build_path: P,
+        path: P,
+        image: &LockedImage,
+        arch: &str,
+    ) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        let tmp_dir = TempDir::new_in(build_path)
+            .context("failed to create a temporary directory for kit extraction")?;
+        let vendor = image.vendor.clone();
+        let name = image.name.clone();
+        let docker_arch = match arch {
+            "x86_64" => Ok("amd64"),
+            "aarch64" => Ok("arm64"),
+            _ => Err(anyhow!("unsupported architecture '{}'", arch)),
+        }?;
+        let target_path = path.as_ref().join(format!("{vendor}/{name}/{arch}"));
+        create_dir_all(&target_path).await?;
+
+        // First get the manifest for the specific requested architecture
+        let manifest = self.get_manifest(image, docker_arch).await?;
+
+        // Now pull the image and extract it to a temporary folder returning
+        // the oci index
+        let digest_file = target_path.join("digest");
+        if digest_file.exists()
+            && read_to_string(&digest_file).await.context(format!(
+                "failed to read cached digest id from {}",
+                digest_file.display()
+            ))? == manifest.digest
+        {
+            return Ok(());
+        }
+
+        let index = self
+            .pull_image(tmp_dir.path(), image, manifest.digest.as_str())
+            .await?;
+
+        // Read the manifest so we can get the layer digests
+        let digest = index
+            .manifests
+            .first()
+            .context("empty oci image")?
+            .digest
+            .replace(':', "/");
+        let manifest_bytes = read(tmp_dir.path().join(format!("blobs/{digest}")))
+            .await
+            .context("failed to read manifest blob")?;
+        let manifest_layout: ManifestLayoutView = serde_json::from_slice(manifest_bytes.as_slice())
+            .context("failed to deserialize oci manifest")?;
+
+        // Extract each layer into the target directory
+        for layer in manifest_layout.layers {
```

Comment:
Doesn't seem like you are cleaning up an existing directory, which could lead to weirdly merged content from two different revisions of the kit.